staging/vhost-device-can/src/vhu_can.rs

```diff
@@ -0,0 +1,2175 @@
+        } else {
+            // If we have a standard frame and can_id > 11 bits then fail
+            if can_id > CAN_SFF_MASK {
+                return Err(Error::UnexpectedCanId(11_u16));
+            }
+        }
+
+        // Remote transfer request is used only with classic CAN
+        if (request.flags.to_native() & VIRTIO_CAN_FLAGS_RTR) != 0 {
+            if !self.check_features(VIRTIO_CAN_F_CAN_CLASSIC)
+                || !self.check_features(VIRTIO_CAN_F_RTR_FRAMES)
+            {
+                warn!("TX: RTR frames not negotiated");
+                return Err(Error::UnexpectedRtrFlag);
+            }
+            can_id |= CAN_RTR_FLAG;
+        }
+
+        // One of VIRTIO_CAN_F_CAN_CLASSIC and VIRTIO_CAN_F_CAN_FD must be negotiated
+        // Check if VIRTIO_CAN_F_CAN_FD is negotiated when the frame is CANFD
+        if (request.flags.to_native() & VIRTIO_CAN_FLAGS_FD) != 0 {
+            if !self.check_features(VIRTIO_CAN_F_CAN_FD) {
+                warn!("TX: FD frames not negotiated\n");
+                return Err(Error::UnexpectedFdFlag);
+            }
+            flags = CAN_FRMF_TYPE_FD;
+        } else {
+            // Check if VIRTIO_CAN_F_CAN_CLASSIC is negotiated when the frame is CAN
+            if !self.check_features(VIRTIO_CAN_F_CAN_CLASSIC) {
+                warn!("TX: Classic frames not negotiated\n");
+                return Err(Error::UnexpectedClassicFlag);
+            }
+        }
+
+        // Adapt CAN length based on negotiated features
+        if (request.flags.to_native() & VIRTIO_CAN_FLAGS_FD) != 0 {
+            if length > 64 {
+                trace!("Cut sdu_len from {:?} to 64\n", request.length);
+                length = 64;
+            }
+        } else if length > 8 {
+            trace!("Cut sdu_len from {:?} to 8\n", request.length);
+            length = 8;
+        }
+
+        // TODO: Check for undefined bits in frame's flags
+
+        Ok(VirtioCanFrame {
+            msg_type: msg_type.into(),
+            can_id: can_id.into(),
+            length: length.into(),
+            reserved: 0.into(),
+            flags: flags.into(),
+            sdu: request.sdu[0..64].try_into().unwrap(),
+        })
+    }
+
+    fn check_rx_frame(&self, mut response: VirtioCanFrame) -> Result<VirtioCanFrame> {
+        CanController::print_can_frame(response);
+
+        let mut can_rx = VirtioCanFrame {
+            msg_type: VIRTIO_CAN_RX.into(),
+            can_id: response.can_id,
+            length: response.length,
+            reserved: 0.into(),
+            flags: response.flags,
+            sdu: [0; 64],
+        };
+
+        // If we receive an error message check if that's a busoff.
+        // If no just drop the message, otherwise update config and return.
+        if (response.can_id.to_native() & CAN_ERR_FLAG) != 0 {
+            if (response.can_id.to_native() & CAN_ERR_BUSOFF) != 0 {
+                // TODO: Trigger a config_change notification
```

Comment:
So here, we should notify the driver that the configuration has changed, right? How do other rust-vmm devices have handled this?

Reply:
Yes, that's right. I did search of it, it's possible that I missed something but I didn't find any device of the existing ones notifying back the driver when changing the configuration. Although, I have found the relevant message that the vhost-devices should send via the socket "[CONFIG_CHANGE_MSG](https://github.com/rust-vmm/vhost/blob/7554d06ff4dbf3c4d7e5c6ac7cb02f83387eac27/vhost/src/vhost_user/message.rs#L203)" when a config change takes place but I didn't find any function on the backend side to trigger this message transmission.

Reply:
I think we can leave it as a comment and then try to fix it later.

Reply:
Great, let's proceed like that